### PR TITLE
C++: SqlPqxxTainted query searches for sql injections via pqxx connector to postgres

### PR DESCRIPTION
**Pull request by japroc**
Hello!

The main goal of my PR it to add support of pqxx connector.
Right now there are only MyQSL and SQlite support. And it is based solely on function name.
Pqxx sink is a bit more complex, because functions like "exec" may lead to high false positive rate.

I basically reused existing SqlTainted query, and created a custom SqlPqxxTainted query. I think that pqxx support should be implemented in Security.qll, but i seems not so simple to overwrite the logic behind SqlTainted query and it's Taint Configuration. That is why i decides to create separate query. I have also added example file SqlPqxxTainted.c and SqlPqxxTainted.qhelp which almost completely repeats SqlTainted.qhelp.

### CODE REVIEW

**Comment by MathiasVP** [comment]
Hi @japroc,

Thank you for your contribution. We will find a reviewer for your query soon :)

I took a quick and very high-level look at your query, and I think it looks good. One thing you need to change is to give your query a _unique_ `@id` (as the one you've picked currently conflicts with the SqlTainted query). Perhaps `cpp/sql-injection-via-pqxx`?

**Comment by japroc** [comment]
Nice, thanks!

I also want to mention that i dont check namespace. It thought that this operation may be redundant.
You can see that second parameter to predicate `pqxxTransationClassNames` is blank. I can add namespace check by adding `t.getParentScope().(Namespace).getName()` instead of `_`.

**Comment by jbj** [comment]
Thank you for the submission. I expect to review this PR next week.

File: cpp/ql/src/experimental/Security/CWE/CWE-089/SqlPqxxTainted.ql [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ * @tags security
+ *       external/cwe/cwe-089
+ */
+
+import cpp
+import semmle.code.cpp.security.Security
+import semmle.code.cpp.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+predicate pqxxTransationClassNames(string class_name, string namespace) {
```
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-089/SqlPqxxTainted.ql:19]: We use [`camelCalse`](https://github.com/github/codeql/blob/main/docs/ql-style-guide.md#naming) for variables in QL.

File: cpp/ql/src/experimental/Security/CWE/CWE-089/SqlPqxxTainted.ql [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  class_name = "transaction" and namespace = "pqxx"
+  or
+  class_name = "basic_transaction" and namespace = "pqxx"
+  or
+  class_name = "transaction_base" and namespace = "pqxx"
+  or
+  class_name = "work" and namespace = "pqxx"
+}
+
+predicate pqxxConnectionClassNames(string class_name, string namespace) {
```
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-089/SqlPqxxTainted.ql:39]: Same here. `class_name` should be `className`.

File: cpp/ql/src/experimental/Security/CWE/CWE-089/SqlPqxxTainted.qhelp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+<p>The code passes user input as part of a SQL query without escaping special elements. 
+It generates a SQL query to Postgres using <code>sprintf</code>,
+with the user-supplied data directly passed as an argument
+to <code>sprintf</code>. This leaves the code vulnerable to attack by SQL Injection.</p>
+
+</overview>
+<recommendation>
+
+<p>Use a library routine to escape characters in the user-supplied
+string before converting it to SQL. Use esc and quote pqxx library functions.</p>
```
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-089/SqlPqxxTainted.qhelp:15]:
```suggestion
string before converting it to SQL. Use <code>esc</code> and <code>quote</code> pqxx library functions.</p>
```

File: cpp/ql/src/experimental/Security/CWE/CWE-089/SqlPqxxTainted.ql [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    function = "quote"
+    or
+    function = "quote_raw"
+    or
+    function = "quote_name"
+    or
+    function = "quote_table"
+    or
+    function = "esc_like"
+  )
```
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-089/SqlPqxxTainted.ql:106]:
```suggestion
  function = ["esc", "esc_raw", "quote", "quote_raw", "quote_name", "quote_table", "esc_like"]
```

**Review by MathiasVP — COMMENTED** [review]
Thanks for your contribution, @aproc!

The PR looks super good, so I hope it's okay I commented on a few stylistic things. When these are fixed I'll happily merge your contribution :)

File: cpp/ql/src/experimental/Security/CWE/CWE-089/SqlPqxxTainted.qhelp [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+</recommendation>
+<example>
+<sample src="SqlPqxxTainted.c" />
```
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-089/SqlPqxxTainted.qhelp:19]:
```suggestion
<sample src="SqlPqxxTainted.cpp" />
```

**Comment by japroc** [comment]
> Thanks for your contribution, @japroc!
> 
> The PR looks super good, so I hope it's okay I commented on a few stylistic things. When these are fixed I'll happily merge your contribution :)

Hi, @MathiasVP!

Nice! I have made changes suggested by you.
What do you think about checking namespace. I could add `t.getParentScope().(Namespace).getName()` in `getPqxxSqlArgument` function and `isEscapedPqxxArgument` predicate. There are some name collisions possible. If you think it's redundant enhancement, we could skip this step and go on.

Thanks!

**Comment by MathiasVP** [comment]
> What do you think about checking namespace. I could add `t.getParentScope().(Namespace).getName()` in `getPqxxSqlArgument` function and `isEscapedPqxxArgument` predicate. There are some name collisions possible. If you think it's redundant enhancement, we could skip this step and go on.

Good point. I think it's a good idea to check for the namespace. Your solution sounds fine, but you might be able to simplify the implementation if you change the type of `t` in `getPqxxSqlArgument`  from `Type` to `Struct`. Then you can use the `getNamespace()` predicate and avoid dealing with parent scopes.

**Comment by japroc** [comment]
> > What do you think about checking namespace. I could add `t.getParentScope().(Namespace).getName()` in `getPqxxSqlArgument` function and `isEscapedPqxxArgument` predicate. There are some name collisions possible. If you think it's redundant enhancement, we could skip this step and go on.
> 
> Good point. I think it's a good idea to check for the namespace. Your solution sounds fine, but you might be able to simplify the implementation if you change the type of `t` in `getPqxxSqlArgument` from `Type` to `Struct`. Then you can use the `getNamespace()` predicate and avoid dealing with parent scopes.

`Struct` type does not fit this case. Because we lose work which is `typedef transaction<> work;`. But `UserType` which is indirect superclass of `Struct` works good.

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**